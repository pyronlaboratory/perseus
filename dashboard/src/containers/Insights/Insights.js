import React, { useState, useEffect, useRef } from "react";

// dashboard components
import StatisticsCard from "../../components/StatisticsCard/StatisticsCard";
import SummaryPanel from "../../components/SummaryPanel/SummaryPanel";
import DoughnutChart from "../../components/DoughnutChart/DoughnutChart";
import BarChart from "../../components/BarChart/BarChart";

// api hooks
import { getFastestUserToCompleteCourse } from "../../hooks/getFastestUserToCompleteCourse";
import { getAverageTimeToCompleteCourse } from "../../hooks/getAverageTimeToCompleteCourse";
import { getMostAttemptedCourse } from "../../hooks/getMostAttemptedCourse";
import { getLeastAttemptedCourse } from "../../hooks/getLeastAttemptedCourse";
import { getUserAttemptPerMonthDistribution } from "../../hooks/getUserAttemptPerMonthDistribution";
import { getUserAttemptPerCourseDistribution } from "../../hooks/getUserAttemptPerCourseDistribution";
import { getFastestSubmissions } from "../../hooks/getFastestSubmissions";
import { getMostSubmissions } from "../../hooks/getMostSubmissions";
import { getRecentSubmissions } from "../../hooks/getRecentSubmissions";

// material imports
import { makeStyles } from '@material-ui/core/styles';

import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';

// styling container
const useStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
        maxWidth: 1064,
        margin: 'auto',
        overflow: 'hidden',
      
    },
    wrapper: {
      padding: theme.spacing(3),
      textAlign: 'center',
      color: theme.palette.text.secondary,
      height: '90%'
    },
    paper: {
        width: '100%'
    },
    title: {
      width: 310,
      fontSize: 16,
      fontWeight: 500,
      padding: theme.spacing(2, 0)
    }
}));

const Insights = () => {

    const classes = useStyles();

    const [fastestUser, setFastestUser] = useState([]);
    const [avgTime, setAvgTime] = useState(0);
    const [mostAttemptedCourse, setMostAttemptedCourse] = useState([]);
    const [leastAttemptedCourse, setLeastAttemptedCourse] = useState([]);

    const [barData, setBarData] = useState([]);
    const [pieData, setPieData] = useState([]);

    const [fastestSubmissions, setFastestSubmissions] = useState([]);
    const [mostSubmissions, setMostSubmissions] = useState([]);
    const [recentSubmissions, setRecentSubmissions] = useState([]);
    
    const componentIsMounted = useRef(true);

    // fetching fastest user to complete a course
    useEffect(() => {
        getFastestUserToCompleteCourse()
        .then(response => {
            if (componentIsMounted.current) {
                setFastestUser(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching average time to complete a course
    useEffect(() => {
        getAverageTimeToCompleteCourse()
        .then(response => {
            if (componentIsMounted.current) {
                setAvgTime(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching most attempted course
    useEffect(() => {
        getMostAttemptedCourse()
        .then(response => {
            if (componentIsMounted.current) {
                setMostAttemptedCourse(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching least attempted course
    useEffect(() => {
        getLeastAttemptedCourse()
        .then(response => {
            if (componentIsMounted.current) {
                setLeastAttemptedCourse(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching users attempt per month distribution
    useEffect(() => {
        getUserAttemptPerMonthDistribution()
        .then(response => {
            if (componentIsMounted.current) {
                setBarData(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching users attempt per course distribution
    useEffect(() => {
        getUserAttemptPerCourseDistribution()
        .then(response => {
            if (componentIsMounted.current) {
                setPieData(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching users with fastest courses completion time
    useEffect(() => {
        getFastestSubmissions()
        .then(response => {
            if (componentIsMounted.current) {
                setFastestSubmissions(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching users with most courses attempted
    useEffect(() => {
        getMostSubmissions()
        .then(response => {
            if (componentIsMounted.current) {
                setMostSubmissions(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    // fetching recent submissions
    useEffect(() => {
        getRecentSubmissions()
        .then(response => {
            if (componentIsMounted.current) {
                setRecentSubmissions(response);
            }
        })
        .catch(err => {
            console.log(err);
        });
        return () => {
            componentIsMounted.current = false;
        };
    }, []);

    return (
        <div style={{  width: '100%' }} className={classes.root}>
        <Grid container spacing={3}>
        
            {/* Statistics Card */}
            <Grid item xs={12} sm={6} lg={3}>
                <StatisticsCard title={"Fastest user to complete a course"} value={fastestUser.user_lastname + ", " + fastestUser.user_firstname}/>
            </Grid>
            <Grid item xs={12} sm={6} lg={3}>
                <StatisticsCard title={"Average course completion time"} value={avgTime + " days"} />
            </Grid>
            <Grid item xs={12} sm={6} lg={3}>
                <StatisticsCard title={"Most attempted course"} value={mostAttemptedCourse.course_title} />
            </Grid>
            <Grid item xs={12} sm={6} lg={3}>
                <StatisticsCard title={"Least attempted course"} value={leastAttemptedCourse.course_title} />
            </Grid>

            {/* Distribution Charts */}        
            <Grid item xs={12} sm={12} md={8} style={{marginBottom: "10px"}}>
                <Typography className={classes.title} color="textSecondary" gutterBottom>User attempts per month distribution</Typography>
                <Paper className={classes.wrapper}>
                    <BarChart data={barData}/>
                </Paper>
            </Grid>
            <Grid item xs={12} sm={12} md={4} style={{marginBottom: "10px"}}>
                <Typography className={classes.title} color="textSecondary" gutterBottom>User per course distribution</Typography>
                <Paper className={classes.wrapper}>
                    <DoughnutChart data={pieData}/>
                </Paper>
            </Grid>
        
            {/* Recent Activities */}
            <Grid item xs={12} sm={12} md={4}>
                <Typography className={classes.title} color="textSecondary" gutterBottom>Recent submissions</Typography>
                <div className={classes.paper}>
                    {recentSubmissions.map((item) => (
                        <SummaryPanel key={item.id} data={item} titleKey={"course_name"} subHeader={"Submitted at"} />
                    ))}

                </div>
            </Grid>
            <Grid item xs={12} sm={12} md={4}>
                <Typography className={classes.title} color="textSecondary" gutterBottom>Fastest submissions</Typography>
                <div className={classes.paper}>
                    {fastestSubmissions.map((item) => (
                        <SummaryPanel key={item.id} data={item} titleKey={"course_name"} subHeader={"Completion time"} />
                    ))}
                </div>
               
            </Grid>
            <Grid item xs={12} sm={12} md={4}>
                <Typography className={classes.title} color="textSecondary" gutterBottom>Most submissions</Typography>
                <div className={classes.paper}>
                    {mostSubmissions.map((item) => (
                        <SummaryPanel key={item.id} data={item} subHeader={"Total courses attempted"} />
                    ))}

                </div>
            </Grid>


        </Grid>
        
        </div>
    );
};

export default Insights;