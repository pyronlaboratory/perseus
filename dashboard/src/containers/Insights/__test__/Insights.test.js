import React from 'react';
import ReactDOM from 'react-dom';

// react testing
import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';

import Insights from './../Insights';

afterEach(cleanup);

describe('insights container', () => {
    it("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(<Insights />, div);
    })

    it("matches snapshot", () => {
        const tree = renderer.create(<Insights />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})