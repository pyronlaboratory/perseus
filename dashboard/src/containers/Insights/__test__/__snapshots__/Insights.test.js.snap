// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insights container matches snapshot 1`] = `
<div
  className="makeStyles-root-1"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root makeStyles-root-5 MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="statistics-card"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root makeStyles-title-6 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
          >
            Fastest user to complete a course
          </p>
          <h1
            className="MuiTypography-root makeStyles-text-7 MuiTypography-h5"
          >
            undefined, undefined
          </h1>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root makeStyles-root-5 MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="statistics-card"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root makeStyles-title-6 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
          >
            Average course completion time
          </p>
          <h1
            className="MuiTypography-root makeStyles-text-7 MuiTypography-h5"
          >
            0 days
          </h1>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root makeStyles-root-5 MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="statistics-card"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root makeStyles-title-6 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
          >
            Most attempted course
          </p>
          <h1
            className="MuiTypography-root makeStyles-text-7 MuiTypography-h5"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root makeStyles-root-5 MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="statistics-card"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root makeStyles-title-6 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
          >
            Least attempted course
          </p>
          <h1
            className="MuiTypography-root makeStyles-text-7 MuiTypography-h5"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-8"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <p
        className="MuiTypography-root makeStyles-title-4 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        User attempts per month distribution
      </p>
      <div
        className="MuiPaper-root makeStyles-wrapper-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <canvas
          data-testid="canvas"
          height={150}
          onClick={[Function]}
          role="img"
          width={300}
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-4"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <p
        className="MuiTypography-root makeStyles-title-4 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        User per course distribution
      </p>
      <div
        className="MuiPaper-root makeStyles-wrapper-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <canvas
          data-testid="canvas"
          height={150}
          onClick={[Function]}
          role="img"
          width={300}
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-4"
    >
      <p
        className="MuiTypography-root makeStyles-title-4 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        Recent submissions
      </p>
      <div
        className="makeStyles-paper-3"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-4"
    >
      <p
        className="MuiTypography-root makeStyles-title-4 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        Fastest submissions
      </p>
      <div
        className="makeStyles-paper-3"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-4"
    >
      <p
        className="MuiTypography-root makeStyles-title-4 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        Most submissions
      </p>
      <div
        className="makeStyles-paper-3"
      />
    </div>
  </div>
</div>
`;
